benchmarks/nel/scripts/datasets/dataset.py

```diff
@@ -0,0 +1,300 @@
+""" Dataset class. """
+import abc
+import importlib
+import inspect
+import os
+import pickle
+from collections import defaultdict
+from pathlib import Path
+from typing import Tuple, Set, List, Optional, TypeVar, Type, Dict
+
+import numpy
+import spacy
+import tqdm
+import yaml
+from spacy import Language
+from spacy.kb import KnowledgeBase
+from spacy.tokens import Doc, DocBin
+from spacy.training import Example
+
+from . import evaluation
+from .utils import ENTITIES_TYPE, ANNOTATIONS_TYPE
+
+DatasetType = TypeVar('DatasetType', bound='Dataset')
+
+
+class Dataset(abc.ABC):
+    """ Base class for all datasets used in this benchmark. """
+
+    KB_VECTOR_LENGTH = 300
+
+    def __init__(self):
+        """ Initializes new Dataset.
+        """
+
+        self._paths = self.assemble_paths(self.name)
+
+        with open(self._paths["root"] / "configs" / "datasets.yml", "r") as stream:
+            self._options = yaml.safe_load(stream)[self.name]
+
+        self._entities: Optional[ENTITIES_TYPE] = None
+        self._failed_entity_lookups: Optional[Set[str]] = None
+        self._annotations: Optional[ANNOTATIONS_TYPE] = None
+        self._kb: Optional[KnowledgeBase] = None
+        self._nlp_base: Optional[Language] = None
+        self._nlp_best: Optional[Language] = None
+        self._annotated_docs: Optional[List[Doc]] = None
+
+    @staticmethod
+    def assemble_paths(dataset_name: str) -> Dict[str, Path]:
+        """ Assemble paths w.r.t. dataset ID.
+        dataset_name (str): Dataset name.
+        RETURNS (Dict[str, Path]): Dictionary with internal resource name to path.
+        """
+
+        root_path = Path(os.path.abspath(__file__)).parent.parent.parent
+        assets_path = root_path / "assets" / dataset_name
+
+        return {
+            "root": root_path,
+            "assets": assets_path,
+            "nlp_base": root_path / "temp" / dataset_name / "nlp",
+            "nlp_best": root_path / "training" / dataset_name / "model-best",
+            "kb": root_path / "temp" / dataset_name / "kb",
+            "corpora": root_path / "corpora" / dataset_name,
+            "entities": assets_path / "entities.pkl",
+            "failed_entity_lookups": assets_path / "entities_failed_lookups.pkl",
+            "annotations": assets_path / "annotations.pkl"
+        }
+
+    @property
+    def name(self) -> str:
+        """ Returns dataset name.
+        """
+        raise NotImplementedError
+
+    def create_knowledge_base(self, model_name: str, **kwargs) -> None:
+        """ Creates and serializes knowledge base.
+        vectors_model (str): Name of model with word vectors to use.
+        """
+
+        self._nlp_base = spacy.load(model_name, exclude=["tagger", "lemmatizer"])
```

Comment:
What's the rationale for the exclude argument here? If it's for speed it would be best to exclude other components too, right?

Reply:
AFAIK (this is an Emerson snippet) this is only for speed reasons. Will drop other components too.